[Tay Jun Yang] iP

dannytayjy opened the pull request:
### Duke

> "Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/dannytayjy/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Create a task list
- [x] Manage the tasks
- [x] Save your tasks

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

dannytayjy committed:
Implement Level-1

* Update Duke.java to include Greet, Echo and Exit features

dannytayjy committed:
Implement Level-2

* Update Duke.java to include Add and List features

dannytayjy committed:
Implement Level-3

* Add Task.java to support representation of tasks
* Update Duke.java to include Done feature

dannytayjy committed:
Implement Level-4 and A-Enums

* Update Task.java to make it an abstract class that represents a task
* Add TaskType.java enumeration to define different types of tasks
* Add TaskWithDateTime.java to make it an abstract class that represents
a task with date/time
* Add ToDo.java, Deadline.java and Event.java to support representation
of three types of tasks
* Update Duke.java to support tracking of three types of tasks

dannytayjy committed:
Implement A-TextUiTesting

* Update input.txt to include inputs for automated text UI testing
* Update EXPECTED.txt to include expected outputs for automated text
UI testing

dannytayjy committed:
Implement Level-5

* Add DukeException.java to handle any exceptions during program
execution
* Add Ui.java to handle input and output interactions
* Update Duke.java to handle errors via exception handling

dannytayjy committed:
Implement Level-6

* Update Duke.java and Ui.java to include Delete feature

dannytayjy committed:
Implement Level-7

* Add Storage.java to handle file operations
* Update Task.java and TaskWithDateTime.java to include string
representation for save format
* Update Duke.java to save tasks in the hard disk automatically
whenever the task list changes

dannytayjy committed:
Implement Level-8

* Update TaskWithDateTime.java with the logic to process date and
time input with yyyy-mm-dd hhmm format into LocalDate and LocalTime
format for output
* Update Deadline.java and Event.java to replace date and time
information taken directly from the input with the processed date
and time information for output
* Update Duke.java to support the software in understanding date
and time in yyyy-mm-dd hhmm format by making minor changes to the
existing implementation when processing date and time of Deadline
and Event tasks

dannytayjy committed:
Fix Level-7

* Update Duke.java, TaskWithDateTime.java, Deadline.java and
Event.java with minor changes to the existing implementation
when processing date and time of Deadline and Event tasks

dannytayjy committed:
Merge branch 'branch-Level-7'

dannytayjy committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/TaskWithDateTime.java

dannytayjy committed:
Implement A-MoreOOP

Make the code more OOP by extracting out more classes

* Add Command.java (interface), AddCommand.java, AddToDoCommand.java,
AddDeadlineCommand.java, AddEventCommand.java, DeleteCommand.java,
ExitCommand.java, ListCommand.java, MarkCommand.java and
UnmarkCommand.java to handle Command execution
* Add CommandType.java enumeration to define different types of
commands
* Add Parser.java to handle the logic of parsing the user command
* Add TaskList.java to contain the list of tasks and handle operations
of adding, deleting and retrieving tasks
* Update Duke.java, Storage.java and Ui.java with changes to support
the implementation of more OOP
* Update TaskType.java Javadoc comments

dannytayjy committed:
Implement A-Packages

* Organize classes into packages

dannytayjy committed:
Implement A-JUnit

* Add JUnit tests in AddCommandTest.java for AddCommand classes
(AddCommand.java, AddToDoCommand.java, AddDeadlineCommand.java
and AddEventCommand.java)
* Add JUnit tests in ParserTest.java for Parser.java
* Add JUnit tests in TaskTest.java for Task classes (Task.java,
TaskWithDateTime.java, ToDo.java, Deadline.java, Event.java)
* Update Storage.java to include getter methods for directory and
dataFile variables so that the variables can be accessed for JUnit
tests in AddCommandTest.java

dannytayjy committed:
Update Level-8, fix minor bugs and enhance codes

Update Level-8

* Add PrintCommand.java to print deadlines/events occurring on a
specific date
* Update other java files with changes to support the implementation
of filtering tasks which is activated when Print command is used
* Update the existing implementation to allow the execution of Mark,
Unmark and Delete commands on filtered tasks

Fix minor bugs

* Update UnmarkCommand.java to use taskNotDoneMessage instead of
taskDoneMessage
* Update Ui.java to remove additional spacing in taskInListMessage
method
* Update Task.java to make corrections to the wrong Javadoc comments
in setNotDone method

Enhance codes

* Update TaskWithDateTime.java to replace if-else code with ternary
operator for processing time input
* Update Javadoc comments for some java files

dannytayjy committed:
Implement Level-9

* Add FindCommand.java to find tasks by searching for a keyword in
the description
* Update other java files with changes to support the implementation
of filtering tasks which is activated when Find command is used

dannytayjy committed:
Implement A-CodingStandard

* Add ErrorMessage.java enumeration to consolidate and define error
messages as constants for reusability and to adhere to coding standard
* Review all java files and tweak some codes to adhere to coding
standard
* Add Javadoc comments for Parser.java

dannytayjy committed:
Implement A-JavaDoc

* Review Javadoc comments of all java files except Parser.java since
major changes have been made to Parser.java in branch-A-CodingStandard

dannytayjy committed:
Merge branch 'branch-Level-9'

dannytayjy committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/task/TaskList.java

dannytayjy committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/task/TaskList.java

dannytayjy committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle-CheckStyle

dannytayjy committed:
Implement A-Gradle and A-CheckStyle

* Add Gradle support to automate project builds
* Add Checkstyle configuration files to detect coding style violations
* Update java files to adhere to coding standard and pass Checkstyle by
correcting detected coding style violations

dannytayjy committed:
Merge branch 'branch-A-Gradle-CheckStyle'

jessicajacelyn commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,37 @@
+    public String dateTimeInfo() {
+        return "(by: " + super.getDateTimeOutput() + ")";
+    }
+}
```
i like how clean the code is here, good job!

jessicajacelyn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,372 @@
+            commandInfo = commandLineParts[1];
+
+            // If user inputs extra text after "bye" or "list" commands
+            if ((commandType.equals("bye") || commandType.equals("list")) && !commandInfo.equals("")) {
```
i think the expression here can be simplified a little because there are too many nested parantheses

jessicajacelyn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,372 @@
+            storage.saveTasksToFile(tasks);
+        } else {
+            throw new DukeException("Task not found. Please try again!");
+        }
```
the name of the methods here are very clear, really nice!

jessicajacelyn reviewed (commented):
Overall i am amazed by the documentation and cleanliness of this program. nothing much to say about what to change, good work! :)

seksek13 commented on src/main/java/TaskWithDateTime.java:
```diff
@@ -1,8 +1,21 @@
  */
 public abstract class TaskWithDateTime extends Task {
     private String dateTimeInput;
+    private String dateInput;
```
I like that your variables names are straightforward and easy to understand! Good job!

seksek13 commented on src/main/java/Duke.java:
```diff
@@ -24,15 +33,15 @@ public void run() {
             }
         }
     }
 
```
Maybe can add javadocs for theses function

seksek13 commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,133 @@
+ * A class that handles create, read and write file operations.
+ */
+public class Storage {
+    private final File dataFile;
```
Shouldn't the constant variables name be all uppercase using underscore to separate words?

seksek13 commented on src/main/java/TaskWithDateTime.java:
```diff
@@ -55,6 +54,15 @@ protected String getDateTime() {
+    public String saveFormat() {
+        return super.saveFormat() + " | " + dateTimeInput;
     }
 }
```
Methods and variables are clean, easing others to understand the code :)

seksek13 reviewed (commented):
Overall coding standards are met for almost all the files, keep it up! 👍

imrajsinghsandhu commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,372 @@
+        }
+    }
+
+    private String processTaskInfoFromDeadlineCommand(String commandInfo) throws DukeException {
+        if (!isCommandInfoPresent(commandInfo)) {
+            throw new DukeException("INCOMPLETE COMMAND"
+                    + System.lineSeparator() + "\t"
+                    + "The description and date/time of a deadline cannot be empty!"
+                    + System.lineSeparator() + "\t"
+                    + "[Note: Enter /by before specifying the date/time]");
+        } else {
+            String[] taskInfoParts = commandInfo.split("/", 2);
+
+            if (taskInfoParts.length != 2) {
+                throw new DukeException("INCOMPLETE COMMAND"
+                        + System.lineSeparator() + "\t"
+                        + "The date/time of a deadline cannot be empty!"
+                        + System.lineSeparator() + "\t"
+                        + "[Note: Enter /by before specifying the date/time]");
+            } else {
+                return getTaskInfoFromDeadlineCommand(commandInfo, taskInfoParts);
+            }
+        }
+    }
```
Generally, your code is very well indented and its done with good coding standards. However, some of your methods are missing javadoc comments, and it would be good to add more!

imrajsinghsandhu commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,133 @@
+import java.io.*;
```
Just a small issue, a wildcard import.

imrajsinghsandhu commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,76 @@
+        return "[" + type.getAbbreviation() + "] " + getStatusIcon() + " " + description;
+    }
+
+    /**
```
This is a great way to do things, and I will try to implement a similar method in my code, as I've done mine in a much more complicated way. Good to see!

imrajsinghsandhu commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,166 @@
+    public String tasksInListMessage(List<Task> tasks) {
+        StringBuilder sb = new StringBuilder();
+
+        sb.append("\t")
+                .append(tasks.size() > 1 ? "Here are the tasks in your list:" : "Here is the task in your list:")
+                .append(System.lineSeparator())
+                .append("\t")
+                .append("[Legend: T = todo, D = deadline, E = event]")
+                .append(System.lineSeparator());
+
```
Well done indentations, good job!

imrajsinghsandhu reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where the nesting was too deep. LGTM!

dannytayjy committed:
Implement Level-10

* Add GUI to Duke using JavaFX technology
* Update other java files and tweak some codes to support the
implementation of GUI

dannytayjy committed:
Merge branch 'branch-Level-10'

dannytayjy committed:
Implement A-Assertions

Add assertions via the assert feature to Ui.java and Command classes
(AddToDoCommand.java, AddDeadlineCommand.java, AddEventCommand.java,
DeleteCommand.java, MarkCommand.java, UnmarkCommand.java,
PrintCommand.java, FindCommand.java).

Assertions allows program to terminate in the case of a possible bug,
and the sooner the execution stops, the safer it is.

Multiple checks for null objects or empty strings are done for the
parameters passed into the methods of Ui.java and the string response
returned from command executions.

Checking for null objects or empty strings will allow us to verify and
ensure that the program is behaving as expected.

dannytayjy committed:
Implement A-CodeQuality

Code quality was improved at the same time when codes were tweaked to
adhere to coding standard.

Nevertheless, code quality is reviewed again for all java files and
some minor changes are made.

dannytayjy committed:
Merge pull request #2 from dannytayjy/branch-A-Assertions

Implement A-Assertions

dannytayjy committed:
Merge pull request #3 from dannytayjy/branch-A-CodeQuality

Implement A-CodeQuality

dannytayjy committed:
Implement C-Help

Add Help Command to provide in-app guidance to users with the available
commands and the respective formats.

dannytayjy committed:
Merge branch 'branch-C-Help'

dannytayjy committed:
Update README.md

dannytayjy committed:
Update README.md

dannytayjy committed:
Correct minor bugs

dannytayjy committed:
Correct minor bugs
